// @abstract-foundation/agw-client/dist/esm/actions/sendPrivyTransaction.js
import { toHex, } from 'viem';
import {} from 'viem/zksync';
import { replaceBigInts } from '../replaceBigInts.js';
export async function sendPrivyTransaction(client, parameters) {
    const result = (await client.request({
        method: 'privy_sendSmartWalletTx',
        params: [replaceBigInts(parameters, toHex)],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }, { retryCount: 0 }));
    return result;
}
export async function sendPrivySignMessage(client, parameters) {
    const result = (await client.request({
        method: 'privy_signSmartWalletMessage',
        params: [parameters.message],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }, { retryCount: 0 }));
    return result;
}
export async function sendPrivySignTypedData(client, parameters) {
    const result = (await client.request({
        method: 'privy_signSmartWalletTypedData',
        params: [client.account.address, parameters],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }, { retryCount: 0 }));
    return result;
}
export async function signPrivyTransaction(client, parameters) {
    const { chain: _chain, account: _account, ...request } = parameters;
    const result = (await client.request({
        method: 'privy_signSmartWalletTx',
        params: [replaceBigInts(request, toHex)],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }, { retryCount: 0 }));
    return result;
}
