// @abstract-foundation/agw-client/dist/cjs/actions/revokeSessions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.revokeSessions = revokeSessions;
const actions_1 = require("viem/actions");
const utils_1 = require("viem/utils");
const SessionKeyValidator_js_1 = require("../abis/SessionKeyValidator.js");
const constants_js_1 = require("../constants.js");
const sessions_js_1 = require("../sessions.js");
async function revokeSessions(client, args) {
    const { session, ...rest } = args;
    const sessionHashes = typeof session === 'string'
        ? [session]
        : Array.isArray(session)
            ? session.map(sessionHash)
            : [(0, sessions_js_1.getSessionHash)(session)];
    const transactionHash = await (0, utils_1.getAction)(client, actions_1.writeContract, 'writeContract')({
        address: constants_js_1.SESSION_KEY_VALIDATOR_ADDRESS,
        abi: SessionKeyValidator_js_1.SessionKeyValidatorAbi,
        functionName: 'revokeKeys',
        args: [sessionHashes],
        ...rest,
    });
    return { transactionHash };
}
function sessionHash(session) {
    if (typeof session === 'string') {
        return session;
    }
    return (0, sessions_js_1.getSessionHash)(session);
}
