// @abstract-foundation/agw-client/dist/cjs/actions/sendPrivyTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendPrivyTransaction = sendPrivyTransaction;
exports.sendPrivySignMessage = sendPrivySignMessage;
exports.sendPrivySignTypedData = sendPrivySignTypedData;
exports.signPrivyTransaction = signPrivyTransaction;
const viem_1 = require("viem");
const replaceBigInts_js_1 = require("../replaceBigInts.js");
async function sendPrivyTransaction(client, parameters) {
    const result = (await client.request({
        method: 'privy_sendSmartWalletTx',
        params: [(0, replaceBigInts_js_1.replaceBigInts)(parameters, viem_1.toHex)],
    }, { retryCount: 0 }));
    return result;
}
async function sendPrivySignMessage(client, parameters) {
    const result = (await client.request({
        method: 'privy_signSmartWalletMessage',
        params: [parameters.message],
    }, { retryCount: 0 }));
    return result;
}
async function sendPrivySignTypedData(client, parameters) {
    const result = (await client.request({
        method: 'privy_signSmartWalletTypedData',
        params: [client.account.address, parameters],
    }, { retryCount: 0 }));
    return result;
}
async function signPrivyTransaction(client, parameters) {
    const { chain: _chain, account: _account, ...request } = parameters;
    const result = (await client.request({
        method: 'privy_signSmartWalletTx',
        params: [(0, replaceBigInts_js_1.replaceBigInts)(request, viem_1.toHex)],
    }, { retryCount: 0 }));
    return result;
}
