// @abstract-foundation/agw-client/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SMART_ACCOUNT_FACTORY_ADDRESS = exports.SESSION_KEY_VALIDATOR_ADDRESS = exports.SESSION_KEY_POLICY_REGISTRY_ADDRESS = exports.NON_EXPIRING_DELEGATION_RIGHTS = exports.INSUFFICIENT_BALANCE_SELECTOR = exports.FEATURE_FLAG_REGISTRY_ADDRESS = exports.EOA_VALIDATOR_ADDRESS = exports.CONTRACT_DEPLOYER_ADDRESS = exports.CANONICAL_EXCLUSIVE_DELEGATE_RESOLVER_ADDRESS = exports.CANONICAL_DELEGATE_REGISTRY_ADDRESS = exports.BRIDGEHUB_ADDRESS = exports.BASE_GAS_PER_PUBDATA_BYTE = exports.AGW_REGISTRY_ADDRESS = exports.AGW_LINK_DELEGATION_RIGHTS = void 0;
const chains_1 = require("viem/chains");
const SMART_ACCOUNT_FACTORY_ADDRESS = '0x9B947df68D35281C972511B3E7BC875926f26C1A';
exports.SMART_ACCOUNT_FACTORY_ADDRESS = SMART_ACCOUNT_FACTORY_ADDRESS;
const EOA_VALIDATOR_ADDRESS = '0x74b9ae28EC45E3FA11533c7954752597C3De3e7A';
exports.EOA_VALIDATOR_ADDRESS = EOA_VALIDATOR_ADDRESS;
const SESSION_KEY_VALIDATOR_ADDRESS = '0x34ca1501FAE231cC2ebc995CE013Dbe882d7d081';
exports.SESSION_KEY_VALIDATOR_ADDRESS = SESSION_KEY_VALIDATOR_ADDRESS;
const CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
exports.CONTRACT_DEPLOYER_ADDRESS = CONTRACT_DEPLOYER_ADDRESS;
const AGW_REGISTRY_ADDRESS = '0xd5E3efDA6bB5aB545cc2358796E96D9033496Dda';
exports.AGW_REGISTRY_ADDRESS = AGW_REGISTRY_ADDRESS;
const INSUFFICIENT_BALANCE_SELECTOR = '0xe7931438';
exports.INSUFFICIENT_BALANCE_SELECTOR = INSUFFICIENT_BALANCE_SELECTOR;
const CANONICAL_DELEGATE_REGISTRY_ADDRESS = '0x0000000059A24EB229eED07Ac44229DB56C5d797';
exports.CANONICAL_DELEGATE_REGISTRY_ADDRESS = CANONICAL_DELEGATE_REGISTRY_ADDRESS;
const CANONICAL_EXCLUSIVE_DELEGATE_RESOLVER_ADDRESS = '0x0000000078CC4Cc1C14E27c0fa35ED6E5E58825D';
exports.CANONICAL_EXCLUSIVE_DELEGATE_RESOLVER_ADDRESS = CANONICAL_EXCLUSIVE_DELEGATE_RESOLVER_ADDRESS;
const AGW_LINK_DELEGATION_RIGHTS = '0xc10dcfe266c1f71ef476efbd3223555750dc271e4115626b';
exports.AGW_LINK_DELEGATION_RIGHTS = AGW_LINK_DELEGATION_RIGHTS;
const NON_EXPIRING_DELEGATION_RIGHTS = `${AGW_LINK_DELEGATION_RIGHTS}000000ffffffffff`;
exports.NON_EXPIRING_DELEGATION_RIGHTS = NON_EXPIRING_DELEGATION_RIGHTS;
const BASE_GAS_PER_PUBDATA_BYTE = 800n;
exports.BASE_GAS_PER_PUBDATA_BYTE = BASE_GAS_PER_PUBDATA_BYTE;
const BRIDGEHUB_ADDRESS = {
    [chains_1.abstractTestnet.id]: '0x35A54c8C757806eB6820629bc82d90E056394C92',
    [chains_1.abstract.id]: '0x303a465b659cbb0ab36ee643ea362c509eeb5213',
};
exports.BRIDGEHUB_ADDRESS = BRIDGEHUB_ADDRESS;
const SESSION_KEY_POLICY_REGISTRY_ADDRESS = '0xfD20b9d7A406e2C4f5D6Df71ABE3Ee48B2EccC9F';
exports.SESSION_KEY_POLICY_REGISTRY_ADDRESS = SESSION_KEY_POLICY_REGISTRY_ADDRESS;
const FEATURE_FLAG_REGISTRY_ADDRESS = '0xb5023a9F3e948e3A4f9DBA97118EEE801fA4e265';
exports.FEATURE_FLAG_REGISTRY_ADDRESS = FEATURE_FLAG_REGISTRY_ADDRESS;
